### PR TITLE
feat:project1 test uint pass

### PR DESCRIPTION
**Pull request by csh0101**
standalone storage coding finished

### CODE REVIEW

File: kv/server/raw_api.go [hunk 1]
```diff
@@ -10,27 +12,99 @@ import (
 
 // RawGet return the corresponding Get response based on RawGetRequest's CF and Key fields
 func (server *Server) RawGet(_ context.Context, req *kvrpcpb.RawGetRequest) (*kvrpcpb.RawGetResponse, error) {
-	// Your Code Here (1).
-	return nil, nil
+	//
```
Comment (BiteTheDDDDt) [inline kv/server/raw_api.go]: remove those `//`

File: kv/server/raw_api.go [hunk 2]
```diff
@@ -10,27 +12,99 @@ import (
...
+	NotFoundResponese := &kvrpcpb.RawGetResponse{
+		NotFound: true,
+	}
+	//
+	reader, err := server.storage.Reader(req.Context)
+	if err != nil {
+		return NotFoundResponese, nil
+	}
+	defer reader.Close()
+	b, err := reader.GetCF(req.Cf, req.Key)
```
Comment (BiteTheDDDDt) [inline kv/server/raw_api.go]: rename `b`, i don't know what it mean.

File: kv/server/raw_api.go [hunk 3]
```diff
@@ -10,27 +12,99 @@ import (
...
+		return &kvrpcpb.RawPutResponse{
+			Error: err.Error(),
+		}, err
+	}
+	return &kvrpcpb.RawPutResponse{}, nil
 }
 
 // RawDelete delete the target data from storage and returns the corresponding response
 func (server *Server) RawDelete(_ context.Context, req *kvrpcpb.RawDeleteRequest) (*kvrpcpb.RawDeleteResponse, error) {
 	// Your Code Here (1).
```
Comment (BiteTheDDDDt) [inline kv/server/raw_api.go]: remove `// Your Code Here (1).`

File: kv/server/raw_api.go [hunk 4]
```diff
@@ -10,27 +12,99 @@ import (
...
+				Key: req.GetKey(),
+			},
+		},
+	})
+	if err != nil {
+		return &kvrpcpb.RawDeleteResponse{
+			Error: err.Error(),
+		}, err
+	}
+	//
```
Comment (BiteTheDDDDt) [inline kv/server/raw_api.go]: same problems

File: kv/server/raw_api.go [hunk 5]
```diff
@@ -10,27 +12,99 @@ import (
...
+	return &kvrpcpb.RawDeleteResponse{}, nil
 }
 
 // RawScan scan the data starting from the start key up to limit. and return the corresponding result
 func (server *Server) RawScan(_ context.Context, req *kvrpcpb.RawScanRequest) (*kvrpcpb.RawScanResponse, error) {
-	// Your Code Here (1).
-	// Hint: Consider using reader.IterCF
-	return nil, nil
+	limit := req.Limit
+	_ = limit
```
Comment (BiteTheDDDDt) [inline kv/server/raw_api.go]: Why we should modify `_`?

File: kv/server/server_test.go [hunk 6]
```diff
@@ -54,7 +54,6 @@ func TestRawGet1(t *testing.T) {
 	server := NewServer(s)
 	defer cleanUpTestData(conf)
 	defer s.Stop()
-
```
Comment (BiteTheDDDDt) [inline kv/server/server_test.go]: do not modify this

File: kv/storage/standalone_storage/standalone_storage.go [hunk 7]
```diff
@@ -1,38 +1,107 @@
...
 	"github.com/pingcap-incubator/tinykv/kv/config"
 	"github.com/pingcap-incubator/tinykv/kv/storage"
+	"github.com/pingcap-incubator/tinykv/kv/util/engine_util"
 	"github.com/pingcap-incubator/tinykv/proto/pkg/kvrpcpb"
 )
 
 // StandAloneStorage is an implementation of `Storage` for a single-node TinyKV instance. It does not
 // communicate with other nodes and all data is stored locally.
 type StandAloneStorage struct {
 	// Your Data Here (1).
```
Comment (BiteTheDDDDt) [inline kv/storage/standalone_storage/standalone_storage.go]: remove all of line like `// Your Data Here (1).`

File: kv/storage/standalone_storage/standalone_storage.go [hunk 8]
```diff
@@ -1,38 +1,107 @@
...
+	//storage engine
+	db     *badger.DB
+	config *config.Config
+	once   *sync.Once
 }
 
 func NewStandAloneStorage(conf *config.Config) *StandAloneStorage {
 	// Your Code Here (1).
-	return nil
+	//new a StandAloneStorage DB
```
Comment (BiteTheDDDDt) [inline kv/storage/standalone_storage/standalone_storage.go]: This comment doesn't look useful

File: kv/storage/standalone_storage/standalone_storage.go [hunk 9]
```diff
@@ -1,38 +1,107 @@
...
+	return &StandAloneStorage{
+		config: conf,
+		once:   &sync.Once{},
+	}
 }
 
+//Start
```
Comment (BiteTheDDDDt) [inline kv/storage/standalone_storage/standalone_storage.go]: same problems

File: kv/storage/standalone_storage/standalone_storage.go [hunk 10]
```diff
@@ -1,38 +1,107 @@
...
 func (s *StandAloneStorage) Start() error {
```
Comment (BiteTheDDDDt) [inline kv/storage/standalone_storage/standalone_storage.go:30]: rename `s`

File: kv/storage/standalone_storage/standalone_storage.go [hunk 11]
```diff
@@ -1,38 +1,107 @@
...
-	// Your Code Here (1).
-	return nil
+	DBPath := s.config.DBPath
+	_, err := os.Stat(DBPath)
+	if err != nil {
```
Comment (BiteTheDDDDt) [inline kv/storage/standalone_storage/standalone_storage.go:36]: add a comment to desctipt this logic.

File: kv/storage/standalone_storage/standalone_storage.go [hunk 12]
```diff
@@ -1,38 +1,107 @@
...
+		db: s.db,
+	}, nil
 }
 
 func (s *StandAloneStorage) Write(ctx *kvrpcpb.Context, batch []storage.Modify) error {
 	// Your Code Here (1).
-	return nil
+	wb := &engine_util.WriteBatch{}
+	for _, v := range batch {
+		switch v.Value() {
```
Comment (BiteTheDDDDt) [inline kv/storage/standalone_storage/standalone_storage.go]: Why we not just use `if`?

File: kv/storage/standalone_storage/standalone_storage.go [hunk 13]
```diff
@@ -1,38 +1,107 @@
...
+			wb.DeleteCF(v.Cf(), v.Key())
+		default:
+			wb.SetCF(v.Cf(), v.Key(), v.Value())
+		}
+	}
+	err := wb.WriteToDB(s.db)
+	return err
+}
+
+var _ storage.StorageReader = new(StandAloneStorageReader)
```
Comment (BiteTheDDDDt) [inline kv/storage/standalone_storage/standalone_storage.go:82]: why define a reader here?

File: kv/storage/standalone_storage/standalone_storage.go [hunk 14]
```diff
@@ -1,38 +1,107 @@
...
+		if err.Error() == badger.ErrKeyNotFound.Error() {
+			return nil, nil
+		}
+		return nil, err
+	}
+	return val, nil
+}
+
+//IterCF
+func (s *StandAloneStorageReader) IterCF(cf string) engine_util.DBIterator {
```
Comment (BiteTheDDDDt) [inline kv/storage/standalone_storage/standalone_storage.go:102]: rename it to some like `GetIterCF`

File: kv/storage/standalone_storage/standalone_storage.go [hunk 15]
```diff
@@ -1,38 +1,107 @@
...
+	txn := s.db.NewTransaction(false)
+	return engine_util.NewCFIterator(cf, txn)
```
Comment (BiteTheDDDDt) [inline kv/storage/standalone_storage/standalone_storage.go]: Do we need to consider when txn is discarded?

File: kv/storage/standalone_storage/standalone_storage.go [hunk 16]
```diff
@@ -1,38 +1,107 @@
...
+}
+
+//Close
+func (s *StandAloneStorageReader) Close() {
```
Comment (BiteTheDDDDt) [inline kv/storage/standalone_storage/standalone_storage.go:110]: add a comment to descript why reader.close do nothing

File: kv/storage/standalone_storage/standalone_storage.go (hunk 14 above)
Comment (csh0101) [reply kv/storage/standalone_storage/standalone_storage.go:102]: the tinykv interface is IterCF

File: kv/storage/standalone_storage/standalone_storage.go (hunk 15 above)
Comment (csh0101) [reply kv/storage/standalone_storage/standalone_storage.go]: Maybe reader txn do not need？ i m not sure

File: kv/storage/standalone_storage/standalone_storage.go (hunk 16 above)
Comment (csh0101) [reply kv/storage/standalone_storage/standalone_storage.go:110]: maybe need to close txn for the reader. but i wonder the relation between  Reader to a Txn is one by one ?

File: kv/storage/standalone_storage/standalone_storage.go (hunk 13 above)
Comment (csh0101) [reply kv/storage/standalone_storage/standalone_storage.go:82]: check the interface implement

File: kv/storage/standalone_storage/standalone_storage.go (hunk 12 above)
Comment (csh0101) [reply kv/storage/standalone_storage/standalone_storage.go]: also work

File: kv/storage/standalone_storage/standalone_storage.go (hunk 11 above)
Comment (csh0101) [reply kv/storage/standalone_storage/standalone_storage.go:36]: sorry it need fix

**Review by BiteTheDDDDt — APPROVED** [review]
LGTM

**Pull request merged**